### DISCUSSION TITLE
Private Inference of MNIST Model

### DISCUSSION THREAD

**Discussion by kanav99** [NillionNetwork/discussions#60] [Show and tell]
### Show and tell project type

Builder Bounty Submission

### Github Repo Link

https://github.com/kanav99/private-mnist

### Video Walkthough Link

https://vimeo.com/963472041?share=copy

### Project Description

In this project, I perform Private Inference of a machine learning model trained over the MNIST dataset, to carry out the task of digit detection. In the process, neither the model owner needs to reveal the model weights, nor the client needs to reveal their input in clear.

### What problems does your project solve? How does it preserve privacy for users?

Machine learning models are usually a proprietary property of companies. On the other hand, inputs to these models might contain some very private data like address, signature, photographs or proprietary source code. Sharing either of the two inputs is a sever breach of privacy. This project solves this using MPC provided by Nillion. 

While in this example, we perform inference of a very simple task of digit detection, we notice that this task comes up in many applications. Every time you convert your hand drawn characters on an iPad, it performs a (possibly remote) inference. It also comes up in optical character/mark recognition (OCR/OMR), which is used in checking exams, reading cheques, etc. Inputs in all of these applications are sensitive, and private inference becomes a necessity in this age.

### How does the project use Nillion? Describe and link to any Nada programs

We use Nada language to describe our model and perform inference privately. You can find the nada program here: https://github.com/kanav99/private-mnist/blob/main/nada/mnist/src/main.py

### Is there anything else you want to share?

The model used in this demo is a simple linear model, which calculates Wx + b, for trained weights W and bias b, followed by an argmax. It looks like when you move to bigger models, like LeNet or MNIST-B, the compilation time is too large - on my laptop, I was not able to compile even in 2hrs. Would love to see optimisations on the compiler side. Natively provided vectori operations would also be really helpful for such ML tasks.

### Optional - Link your project and team members' social handles

https://x.com/kanavgupta99

### Optional - Team ETH Address(es)

- 0x91CD3caDe0BFd95Ad1d37BD058E05B04B4181A44

**Comment by mathias-nillion** [comment]
Hi @kanav99, thanks for contributing this project - it looks awesome!
We heard your feedback about improving data & AI workflows loud and clear.
This is why we made https://github.com/NillionNetwork/nada-numpy (improved data workflows) and https://github.com/NillionNetwork/nada-ai (improved AI workflows) to help AI builders such as yourself continue to build cool stuff on Nillion :rocket:
Feel free to check it out & let us know what you think!

PS: as for compiler performance, we're cooking up some nice improvements here - stay tuned!

**Reply by kanav99** [reply]
Hey @mathias-nillion, thank you so much for working on this! Was a perfect timing. I just tried and now I was able to evaluate a larger model and it works! both compilation and running. Concretely, I was able to run a 784x128, 128x10 model for MNIST inference. 

Kudos to the entire team for building such a usable set of tools!

**Reply by mathias-nillion** [reply]
Hi @kanav99, great to hear that - appreciate it!
Feel free to let us know if you have suggestions on how we can support you further 🚀

**Comment by oceans404** [comment]
Congratulations @kanav99!!! You won our tinybin bounty. Can you confirm that the ETH address you provided is correct so we can send your prize: 0x91CD3caDe0BFd95Ad1d37BD058E05B04B4181A44

tinydemo winners announcement tweet: https://x.com/nillionnetwork/status/1820606184895947087

**Comment by kanav99** [comment]
Thank you so much @oceans404! It's an honour! Yes, I confirm that this is my ETH address

**Reply by jimouris** [reply]
Congrats @kanav99!

**Reply by Nillion-Engineering** [reply]
Hey, Claire from the Nillion Association here! Thanks for building on Nillion. Here's the link to your prize. Well done!

https://etherscan.io/tx/0x029b28bd884643a41b61bf2883435aaa97bd0e8c3dd1d269d4b1ab2e1d05c214

**Reply by kanav99** [reply]
Thank you so much Claire, I have received the prize!

**Comment by kanav99** [comment]
Hey @oceans404, was the transfer made? Thanks!

**Reply by oceans404** [reply]
Hey @kanav99, no we haven't made the transfer yet, but I'll send the transaction hash from our finance team as soon as we make it
